algorithmic-strategies/problem-set-algorithmic-strategies.md

```diff
@@ -0,0 +1,336 @@
+        [104, 105, 138, 135, 82, 73, 103, 52, 57, 136],
+        75
+    ) == 188
+
+def test_large_knapsack(self):
+    assert knapsack(
+        [92, 39, 56, 81, 59, 38, 71, 91, 73, 34, 13, 27, 10, 49, 56, 90, 66, 91, 47, 52, 46, 43, 69, 53, 90, 99, 84, 41, 33, 89, 46, 78, 29, 16, 58, 44, 11, 87, 26, 13, 11, 63, 81, 51, 12, 90, 29, 100, 93, 43, 52, 65, 45, 46, 45, 98, 74, 58, 60, 100, 42, 66, 56, 37, 80, 68, 91, 63, 65, 73, 59, 10, 41, 30, 88, 79, 11, 94, 20, 89, 83, 70, 47, 75, 19, 15, 33, 34, 94, 65, 66, 44, 62, 95, 45, 17, 92, 93, 27, 43],
+        [96, 56, 143, 123, 96, 83, 82, 117, 82, 52, 66, 126, 122, 90, 128, 149, 149, 116, 55, 145, 111, 145, 63, 137, 85, 148, 70, 125, 57, 72, 82, 124, 107, 137, 127, 139, 125, 94, 111, 108, 117, 142, 121, 103, 140, 54, 82, 55, 104, 50, 98, 82, 50, 64, 105, 112, 70, 113, 91, 72, 108, 112, 132, 141, 85, 138, 94, 117, 137, 104, 54, 135, 105, 95, 135, 150, 143, 144, 143, 132, 52, 90, 100, 119, 59, 147, 125, 55, 101, 90, 72, 90, 58, 131, 123, 76, 109, 130, 138, 77],
+        200
+    ) == 1657
+```
+
+### !end-question
+### !placeholder
+
+```python
+def knapsack(weights, values, max_weight):
+    pass
+```
+### !end-placeholder
+### !tests
+```python
+import unittest
+from main import *
+
+class TestZeroOneKnapsack(unittest.TestCase):
+    def test_empty_knapsack(self):
+        self.assertEqual(knapsack([], [], 0), 0)
+
+    def test_small_knapsack(self):
+        self.assertEqual(knapsack([10, 20, 30], [60, 100, 120], 50), 220)
+
+    def test_medium_knapsack(self):
+        self.assertEqual(knapsack([99, 60, 70, 68, 50, 65, 48, 31, 52, 29], [104, 105, 138, 135, 82, 73, 103, 52, 57, 136], 75), 188)
+
+    def test_large_knapsack(self):
+        self.assertEqual(knapsack([92, 39, 56, 81, 59, 38, 71, 91, 73, 34, 13, 27, 10, 49, 56, 90, 66, 91, 47, 52, 46, 43, 69, 53, 90, 99, 84, 41, 33, 89, 46, 78, 29, 16, 58, 44, 11, 87, 26, 13, 11, 63, 81, 51, 12, 90, 29, 100, 93, 43, 52, 65, 45, 46, 45, 98, 74, 58, 60, 100, 42, 66, 56, 37, 80, 68, 91, 63, 65, 73, 59, 10, 41, 30, 88, 79, 11, 94, 20, 89, 83, 70, 47, 75, 19, 15, 33, 34, 94, 65, 66, 44, 62, 95, 45, 17, 92, 93, 27, 43], [96, 56, 143, 123, 96, 83, 82, 117, 82, 52, 66, 126, 122, 90, 128, 149, 149, 116, 55, 145, 111, 145, 63, 137, 85, 148, 70, 125, 57, 72, 82, 124, 107, 137, 127, 139, 125, 94, 111, 108, 117, 142, 121, 103, 140, 54, 82, 55, 104, 50, 98, 82, 50, 64, 105, 112, 70, 113, 91, 72, 108, 112, 132, 141, 85, 138, 94, 117, 137, 104, 54, 135, 105, 95, 135, 150, 143, 144, 143, 132, 52, 90, 100, 119, 59, 147, 125, 55, 101, 90, 72, 90, 58, 131, 123, 76, 109, 130, 138, 77], 200), 1657)
+```
+### !end-tests
+### !hint
+
+- We probably need to add an additional parameter to be able to store and pass memoized data.
+- What data type would be appropriate for storing the memoized data?
+- How can we initialize the storage for the memoized data?
+- Where do we need to update our memoized data?
+- Where do we need to pass our memoized data?
```

Comment:
If you'd like to have mercy on your fellow instructors, it'd be awesome if you could add these questions and small possible answers in the instructor document!

(Every question we pose students is a question they'll want an answer to from an instructor. Also, there will be a roundtable on this homework! :P)

I'll comment to point out all the questions possible, and here's some possible answers...

- What data type would be appropriate for storing the memoized data?
  - Dictionary, because we need to keep the max value and a way to identify the max value
- How can we initialize the storage for the memoized data?
  - `memo = {}` at top of the function, if `memo` doesn't ext
- Where do we need to update our memoized data?
  - In each function after we find a possible max value
- Where do we need to pass our memoized data?
  - Whenever we look for another possible max value (every time we call `knapsack()`, which is 2 or 3 times in most solutions)

Reply:
Adding discussion points for all the hint prompts to the instructor doc.